### PR TITLE
Add tutorial for in-place solve

### LINKED ISSUE #117: Add documentation for in-place optimization

### PR DESCRIPTION
**Pull request by tmigot**
Close #117

### CODE REVIEW

File: docs/src/resolve.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# [Resolve and in-place solve](@id resolve)
```
Comment (dpo) [inline docs/src/resolve.md:1]:
```suggestion
# [Re-solve and in-place solve](@id resolve)
```

File: docs/src/resolve.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+It is very convenient to pre-allocate the memory used during the optimization of a given problem either from improved memory management or when re-solving the same or a similar problem.
+
+Let us consider the following 2-dimensional unconstrained problem
+```math
+\begin{aligned}
+\min \quad & f(x):= x₂ exp(x₁) 
+\end{aligned}
+```
+Using `JSOSuite` usual `minimize` function, the problem can be solved as follows
```
Comment (dpo) [inline docs/src/resolve.md:11]:
```suggestion
Using `JSOSuite`’s `minimize` function, the problem can be solved as follows
```

File: docs/src/resolve.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+```@example ex1
+using JSOSuite
+f(x) = x[2]^2 * exp(x[1]^2)
```
Comment (dpo) [inline docs/src/resolve.md:14]: This isn’t the same function as above.

File: docs/src/resolve.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+x0 = ones(Float64, 2)
+stats = minimize(f, x0)
+```
+Using L-BFGS, the problem is locally solved.
+
+## In-place solve
+
+If we want to solve the same problem several time, for instance, for several initial guess it is recommended to use in-place solve.
```
Comment (dpo) [inline docs/src/resolve.md:22]:
```suggestion
If we want to solve the same problem several times, for instance, for several initial guesses, it is recommended to use an in-place solve.
```

File: docs/src/resolve.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+```@example ex1
+using ADNLPModels, JSOSolvers, SolverCore
+f(x) = x[2]^2 * exp(x[1]^2)
+x0 = ones(Float64, 2)
+nlp = ADNLPModel(f, x0) # or use JuMP
+solver = JSOSolvers.LBFGSSolver(nlp)
```
Comment (dpo) [inline docs/src/resolve.md:28]: Maybe explain that you can’t just pass `f` anymore, but you need to create a model.

File: docs/src/resolve.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+stats = SolverCore.GenericExecutionStats(nlp)
+solve!(solver, nlp, stats, x = x0)
+```
+This deserves more explanation.
```
Comment (dpo) [inline docs/src/resolve.md:32]:
```suggestion
This deserves more explanations.
```

File: docs/src/resolve.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+The name of the solver structure and the corresponding package can be accessed via the DataFrame `JSOSuite.optimizers`.
+```@example ex1
+JSOSuite.optimizers[!, [:name_solver, :name_pkg]]
+```
+In our example, the solver LBFGS is implemented in `JSOSolvers.jl` and the solver structure is `LBFGSSolver`.
+
+Now, it is possible to reuse the memory allocated for the first solve for another round.
```
Comment (dpo) [inline docs/src/resolve.md:39]:
```suggestion
Now, it is possible to reuse the memory allocated for the first solve for another round:
```

File: docs/src/resolve.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+```@example ex1
+# NLPModels.reset!(nlp) would also reset the evaluation counters of the model
+reset!(solver)
+x02 = [1.4; 5.0] # another initial guess
```
Comment (dpo) [inline docs/src/resolve.md:43]:
```suggestion
new_x0 = [1.4; 5.0] # another initial guess
```

File: docs/src/resolve.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+solve!(solver, nlp, stats, x = x02)
```
Comment (dpo) [inline docs/src/resolve.md:44]:
```suggestion
solve!(solver, nlp, stats, x = new_x0)  # new solve with existing solver object
```

File: docs/src/resolve.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+```
+
+## In-place solve of a different problem
+
+It is also possible to reuse the allocated memory to solve another problem with the same characteristics.
```
Comment (dpo) [inline docs/src/resolve.md:49]:
```suggestion
It is also possible to reuse the allocated memory to solve another problem with the same number of variables and constraints:
```
Question: must `nnzh` and `nnzj` also be the same?

File: docs/src/resolve.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+In order to measure, the amount of allocations made by the solvers the package NLPModelsTest.jl defines a set of test problems that are allocation-free.
+```@example ex1
+using NLPModelsTest, SolverCore, JSOSolvers
+nlp = BROWNDEN(Float64)
+solver = LBFGSSolver(nlp)
+stats = GenericExecutionStats(nlp)
+solve!(solver, nlp, stats)
+@allocated solve!(solver, nlp, stats)
+```
+Several of the pure Julia implementation available in JSOSuite have this property.
```
Comment (dpo) [inline docs/src/resolve.md:69]:
```suggestion
Several of the pure Julia solvers available in JSOSuite have this property.
```

**Review by dpo — CHANGES_REQUESTED** [review]

File: docs/make.jl [hunk 12]
```diff
@@ -15,6 +15,7 @@ makedocs(
     "Tutorial" => "tutorial.md",
     "Nonlinear Least Squares" => "nls.md",
     "Quadratic models with linear constraints" => "qp.md",
+    "Resolve and in-place solve" => "resolve.md",
```
Comment (tmigot) [inline docs/make.jl:18]:
```suggestion
    "Re-solve and in-place solve" => "resolve.md",
```

File: docs/src/resolve.md (hunk 5 above)
Comment (tmigot) [reply docs/src/resolve.md:38]: I added a paragraph above to justify further. That would be something great to add too, the `solve!(solver, f, x0, stats; kwargs...)`.

File: docs/src/resolve.md (hunk 10 above)
Comment (tmigot) [reply docs/src/resolve.md:59]: It depends on what the solver needs. If we pre-allocate some space for Jacobian and Hessian values, then yes.

**Review by dpo — CHANGES_REQUESTED** [review]
Can be merged after these changes.

File: docs/src/resolve.md [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+\end{aligned}
+```
+Using `JSOSuite`’s `minimize` function, the problem can be solved as follows
+```@example ex1
+using JSOSuite
+f(x) = x[2]^2 * exp(x[1]^2)
+x0 = ones(Float64, 2)
+stats = minimize(f, x0)
+```
+Using L-BFGS, the problem is locally solved.
```
[the remark this answers is not in the document]
Comment (tmigot) [reply docs/src/resolve.md:18]: I found a way to show the console output :), which justifies it.

**Pull request merged**